### ISSUE TITLE
[RFC] Allow inline variable declaration syntax in `if`, `while` (and `switch`)

### ISSUE THREAD

**Issue by mhermier** [wren-lang/wren#969]
After some discussions in #956, #964 and it has become apparent that there is a need to catch the result of the testing part.

Currently there are 2 possibility that are considered:
```javascript
keyword (var variable_name = expr) ...
```
and
```javascript
keyword (expr) as variable_name ...
```

The first one seems a little bit more natural, and seems to follow what other languages do.

The second has the advantage to label/tag the structure to manage the execution flow without the need to add an extra labeling system.

**Comment by PureFox48** [comment]
I wouldn't be against the first of those options to avoid the repetition in:

```javascript
var variable_name
keyword (variable_name = expr) { ... }
```
It would also have the advantage of limiting the scope of `variable_name` to the `if`, `while` or `switch` block itself which would usually be what you'd want.

A slight drawback is that it wouldn't be consistent with the `for` statement whose control variable doesn't require `var` to achieve these aims.

**Comment by cxw42** [comment]
@mhermier I thought about this and actually started drafting an issue :) .  However, while typing, I had second thoughts about using this anywhere other than `switch`.

- `if(var x = expr)`: Every value in Wren is truthy except two.  Therefore, anywhere I would care about the actual value, and not just its truthiness, I would use `switch` instead of `if`.  
  - I agree with @PureFox48 that restricting the variable's scope is a benefit.  I am having trouble thinking of a situation where the result of the truth test is what I want.  E.g., in C, I might well write `if((foo=expr)>42) {}`, but I don't think I've ever written `if(foo=(expr>42)) {}`.  As far as I can tell, the latter is what `var x = expr` looks like.
- `while(var x = expr)`: since `expr` is evaluated each time through the loop, `x` would logically be re-created/assigned each time through the loop.  If I care about the value of the loop counter, I usually use `for` rather than `while`.
- `for`: as @PureFox48 just pointed out, it already has different syntax.

For `switch`, though, being able to conveniently give the `expr` a name makes a great deal of sense.  There are many possible values, and multiple tests (not just for truthiness).  Therefore, I suggest we keep this discussion in #956 for now.  What say you?

**Comment by cxw42** [comment]
@mhermier Re. option 2, there are at least two kinds of simplicity.  I think you and I disagree on which makes the most sense here.  

- One kind is "use the fewest keywords".  I think that is what you are advocating.  Yes?
- Another kind is "reduce the user's mental load."  I think that separating the expression from the name with a paren increases mental load.  I am used to thinking of parens as boundaries.  If I understand you correctly, `if(expr) as x` forces me to ignore that `)` boundary and mentally attach `expr` and `x`.

I look forward to your thoughts!

**Comment by PureFox48** [comment]
@cxw42
I wondered myself about the continual redeclaration implicit in using this syntax for the `while` statement but I don't think it would matter in practice - the variable would always use the same memory location.

**Comment by cxw42** [comment]
@PureFox48 I agree it should be implementable efficiently.  Can you think of any use cases for this in `while`, though?  I can't.  

The closest I've been able to think of is the iterator use case.  [wren.io](https://wren.io/control-flow.html#the-iterator-protocol) gives this example of how the iterator protocol works under the hood:

```js
var iter_ = null
var seq_ = 1..100
while (iter_ = seq_.iterate(iter_)) {
  var i = seq_.iteratorValue(iter_)
  System.print(i)
}
```
Now, `iter_` doesn't need to exist outside the loop, so it would be a good candidate for a `while(var x=...)`.  But then how do you initialize it?
```
var seq_ = 1..100
while (var iter_ = seq_.iterate(???)) {
  var i = seq_.iteratorValue(iter_)
  System.print(i)
}
```
The `???` has to be something different for the first time through the loop than through subsequent times through the loop.

I think `switch(var x = expr)` is valuable independent of `if`/`while`/`for`.  I realize I opened this can of worms in the first place :) , and I apologize for unintentionally causing confusion!

**Comment by PureFox48** [comment]
I agree that this idea would be most valuable in the context of the`switch` statement and I think it would be worth doing for that reason alone.

I'm not so bothered about `if` and `while` where the only advantage seems to be limit the scope of the variable.

**Comment by mhermier** [comment]
I'm not not advocating for one in particular, both have valid reasons to exist.

The second one as a slight advantage of reducing a little the verbosity in the cases where we need a jump label (and I don't state them because not all the case have been completely determined yet). It is an oddity compared to other languages, but in the same time would enforce the wren `as` syntax.

One valid reason to use first form would be performance. Second form implies that when a label is needed, even if the variable is never used, we store the variable till it can get garbage collected at next loop or loop exit.

`if(var x = expr)` Ease null check writing, and is never resolved with a `switch`.

`while(var x = expr)` unless I'm mistaken, the idiom can be useful for them generator (unless you use an adapter to transform it to a sequence, which would defeat the point of using a generator)

The `for` syntax as a different syntax, this is true. And I don't think we can't really do something about it unfortunately. Thought it would be another good advocate for first form, since `for (var x in exp)` fells natural. (And this is where I was mistaken, `for` is the one that implicitly creates variables... and is a little confusing because it *should* call 'this.x' and 'this.x=(_)' to be coherent)

The iterator example cannot benefit from the syntax (UNLESS we change the semantic of `var` so it `null` initialize/declare the variable before we affect them. It would solve a bunch of problems like the impossibility to make recursive function simply but would open a can of worm around initialization reasoning).

**Comment by ChayimFriedman2** [comment]
Regarding the iterator example, we _can_ make it possible to access a variable in its own initializer, but it will be a breaking change to do that everywhere (previously this referenced shadowed variable, if any), so we'll have to do this only for `while`, and this is confusing.

**Comment by cxw42** [comment]
@mhermier I like your point about null checks.  I therefore support if(var...) and switch(var...).

**Comment by ChayimFriedman2** [comment]
Another option will be to allow Go style statement before conditions:
```wren
if (var a = true; a) { ... }
```

Also we may want to consider multiple variable declarations in one statement (`var a = 5, b = false`), so that whatever syntax we choose you can declare multiple variable.

**Comment by mhermier** [comment]
To confirm this, I suggest that the continuation label is made.

Once accepted, a proof of concept change can be made using the second syntax. It would allow to make a comparison point with a benchmark, and give more proofs to the fact that it is better to do it separately, and use the first syntax for good.

About the iterator example, it was an explanation of why we cannot do it. So for now we should stick to the current semantic, unless it become a visible burden.

**Comment by PureFox48** [comment]
ISTM that this proposal has become inextricably linked with #956 and #964 and that we should therefore wait for @Ruby0x1's ruling on those before trying to progress this further. If it were decided that either or both of those proposals were too complicated  or not worth doing, then we might view this one in a different light.

**Comment by cxw42** [comment]
I support waiting for more on #956 switch.  I don't think this issue needs to wait for #964 multilevel break, personally, but I also don't have any reason to rush on this issue :) .

**Comment by mhermier** [comment]
I did the implementation with `var` for testing and it rise a problem where we get a *surprising* behavior with:
```javascript
var a = expr
while (var a = fn.call(a) ) {
  // `fn.call(a)` will refers to the upper level `a`
  // meaning encapsulated `a` will always have the same value
  // resulting in an infinite loop
  ...
}
```

**Comment by PureFox48** [comment]
If I were you, I wouldn't try and implement this feature for loops.

Just do it for `if` and (potentially) `switch` which won't have the problem of being continually re-declared.

**Comment by mhermier** [comment]
It is not a problem of not doing it, it was trivial ~ 20 lines to add to the compiler, so it is conforming to the `var` semantic. But it has that behavior...

**Comment by PureFox48** [comment]
Well, if you just do it for `if` and `select` then `var a = fn.call(a)` will only be computed once, rather than each time around the loop, so you won't have this problem.

**Comment by cxw42** [comment]
@mhermier thanks for implementing!  I agree that behaviour would be surprising to the Wren programmer and that it should only be `if` for now.

**Comment by mhermier** [comment]
The implementation comes with the revival of `DUP` opcode (why oh why all the good stuff need it XD), so it has to be thought a little bit more first.

I think a better question is: does the `var` semantic still makes sense because of that shadowing feature/issue ?
I mean, the original intent was to avoid some initialization issues. But:
1. It introduce a first surprise because it refers to shadowed variables during initialization (which can be replaced by an up-value lookup operator if really needed in the future).
2. It remove the possibility to easily make recursive functions (which is quite a deal, but can also be solved by adding proper named functions, or the ugly pre-define the variable before setting it to the real value).
3. Makes things not clear (as in 1) when applied to loops.
The only advantage is that it avoid a little of stack manipulation at initialization.

I mean AFAIK shadowing a variable is a bad practice, and I don't see a legitimate case where this could not be a source of confusion, in particular when a mark of upvalue lookup is not present.

**Comment by PureFox48** [comment]
Well, bad practice or not, this feature would need to cope with shadowing because it's allowed everywhere else in the language.

If the results are surprising when it's used with `while` then I think that's telling us something.

**Comment by mhermier** [comment]
I don't reject the arguments, but I think they are only valid if you consider the value as not defined prior to usage. Most languages assume that the bits gets initialized after the assignment. We have the luxury to be able to default null initialize the value before assign it for cheap.

**Comment by cxw42** [comment]
@mhermier would you be willing to open a draft PR or point me to your branch?  I think I will understand your points better if I look at the code.  Thank you!

**Comment by mhermier** [comment]
This is experimental, naming and implementation is not final:
https://github.com/mhermier/wren/tree/experimental_inline_variable_declaration

**Comment by cxw42** [comment]
@mhermier Thank you very much indeed!  The code is very simple, as you said :) .  Per your https://github.com/wren-lang/wren/issues/969#issuecomment-821829837:

> 1. It introduce a first surprise because it refers to shadowed variables during initialization

I just tried this in a different context:
```
$ cat foo.wren 
var a = 42
if(true) {
        var a = a+1      // <=== 
        System.print(a)
}

$ wren_cli foo.wren
43
```

I think the shadowing is OK since it is consistent with the rest of the language (https://github.com/wren-lang/wren/issues/969#issuecomment-821834037)

> 2. It remove the possibility to easily make recursive functions (which is quite a deal, but can also be solved by adding proper named functions, or the ugly pre-define the variable before setting it to the real value).

I looked at the code and I do not understand how this prevents easily making recursive functions.  Could you perhaps show a code example of what doesn't work with your change?  That would be very helpful!

> 3. Makes things not clear (as in 1) when applied to loops.

I agree this should not be in loops.  It is indeed very confusing in a loop condition :) .

**Comment by mhermier** [comment]
The branch I published contains the conformant version, I need to do a little bit more work to make the other behavior.

To understand the problem with functions I'll take the trivial Fibonacci function. One would trivially write it like this:
```javascript
var fib = Fn.new {|n| n < 2 ? n : fib.call(n - 1) + fib.call(n - 2) }
```
That does not work because of the rule, so to have a working solution you can pre-declare the symbol:
```javascript
var fib
fib = Fn.new {|n| n < 2 ? n : fib.call(n - 1) + fib.call(n - 2) }
```
or do it a functor:
```javascript
class Fib {
  static call (n) { n < 2 ? n : call(n - 1) + call(n - 2) }
}
```
There is a possibility that doesn't work is to be able to access `this` but it does not solve the issue when you have a recursive function based on 2+ functions.
```javascript
var fib = Fn.new {|n| n < 2 ? n : this.call(n - 1) + this.call(n - 2) }
```
Of all these functions, the first one seems the more natural too me. Resolving the issue of recursive functions based on 2+functions would require to have first function definition which is a no no based on the principles of the language.

By changing the behavior of `var`, making the variable accessible and initialized (trivially we can choose `null` which sounds reasonable). Suddenly you can write things like this:
```javascript
var fib = Fn.new {|n| n < 2 ? n : fib.call(n - 1) + fib.call(n - 2) }
while (var iter = sequence.iterate(iter)) ...
```
I don't mean it is perfect and might produce its own class of bugs when used in expressions, but it is more interesting than trying to access a to be shadowed variable.

**Comment by ChayimFriedman2** [comment]
There is a reason it is done this way, Bob specified it in Crafting Interpreters: http://craftinginterpreters.com/resolving-and-binding.html#resolving-variable-declarations.

The disadvantage is that functions sometimes _do_ need to access themselves.

This has two solutions:

 - In one of my languages, I've allowed referencing a variable inside its own initializer _if we entered a new closure_. This way, buggy statements like `var a = a` will be rejected, but `var a = Fn.new {|b| b == 0 ? 0 : a.call(b - 1) }` won't. This works beautifully, but I don't think Wren needs that, because...
 - Differentiating accessible-before-declaration and non-accessible-before-declaration variables. This is what Wren does, using the case of the first letter as the differentiator. Classes are no more special than functions - the same way your class `Fib` refers itself, so can do `fib`:
```wren
var Fib = Fn.new {|n| n < 2 ? n : Fib.call(n - 1) + Fib.call(n - 2) }
```

**Comment by mhermier** [comment]
That differentiatior doesn't work as soon as you enter a scope as per definition of that rules. Most of function (in particular the ones with captured values) are not defined at the module level. Excluding the generic(library grade) ones that can leave there, it means that functions are not fully usable.

At the end of the day nothing change, we always come back to that rules, proving it is more annoying than it helps...

**Comment by sukhchainn** [comment]
I vote for `keyword (expr) as variable_name`

It makes the code more obvious than looking up the docs or worse (guessing) for those who are not much familiar with wren.
